lib/sycamore/sycamore/llms/openai.py

```diff
@@ -352,6 +395,63 @@ def _generate_using_openai(self, prompt_kwargs, llm_kwargs) -> str:
         completion = self._client.chat.completions.create(model=self._model_name, messages=messages, **kwargs)
         return completion.choices[0].message.content
 
+    async def generate_future(self, *, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None) -> Awaitable[str]:
+        cache_key = None
+        if self._cache:
+            cache_key = self._get_cache_key(prompt_kwargs, llm_kwargs)
+            hit = self._cache.get(cache_key)
+            if hit:
+                if (
+                    hit.get("prompt_kwargs") == prompt_kwargs
+                    and hit.get("llm_kwargs") == llm_kwargs
+                    and hit.get("model_name") == self.model.name
+                ):
+                    return hit.get("result")
+                else:
+                    logger.warning(
+                        "Found cache content mismatch, key=%s prompt_kwargs=%s llm_kwargs=%s model_name=%s",
+                        cache_key,
+                        prompt_kwargs,
+                        llm_kwargs,
+                        self.model.name,
+                    )
+        if llm_kwargs is not None:
+            result = await self._generate_future_using_openai(prompt_kwargs, llm_kwargs)
+        else:
+            raise ValueError("Must include llm_kwargs to generate future call")
+        if self._cache:
+            assert cache_key
+            item = {
+                "result": result,
+                "prompt_kwargs": prompt_kwargs,
+                "llm_kwargs": llm_kwargs,
+                "model_name": self.model.name,
+            }
+            self._cache.set(cache_key, item)
+        return result
+
+    async def _generate_future_using_openai(self, prompt_kwargs, llm_kwargs) -> Awaitable[str]:
+        kwargs = {
```

Comment:
```
  kwargs = self._get_generate_kwargs(prompt_kwargs, llm_kwargs)
```
include messages in kwargs
and then share with _generate_using_openai

Reply:
changed